### PR TITLE
open unsupported documents as text

### PR DESCRIPTION
**Pull request by yparitcher**
This ~~is a working POC / hack to~~ allow opening unsupported files as txt, for example .lua files.
this can be very helpful when having oddly named files. 
it might be a good idea to allow all providers on this list?

this only triggers via the open with menu, not by directly clicking the item, unless setting the provider as a default.

this probably needs proper integration but i am not 100% sure exactly how the `DocumentRegistry` works and don't want to mess things up.

~~once this works i would like to add a way to clear already set providers from settings.~~
 
FileManager_2019-Oct-17_015418

### CODE REVIEW

File: frontend/apps/filemanager/filemanager.lua [hunk 1]
```diff
@@ -293,8 +293,8 @@ function FileManager:init()
             {
                 {
                     text = _("Open with…"),
-                    enabled = lfs.attributes(file, "mode") == "file" and DocumentRegistry:getProviders(file) ~= nil
-                        and #(DocumentRegistry:getProviders(file)) > 1,
+                    enabled = lfs.attributes(file, "mode") == "file" and (DocumentRegistry:getProviders(file) == nil
+                        or #(DocumentRegistry:getProviders(file)) > 1),
```
Comment (Frenzie) [inline frontend/apps/filemanager/filemanager.lua]: Looks like you want to remove everything after file?

File: frontend/ui/widget/filechooser.lua [hunk 2]
```diff
@@ -395,17 +395,27 @@ function FileChooser:showSetProviderButtons(file, filemanager_instance, reader_u
...
+                },
+            })
+        end
+    else
+        local provider = DocumentRegistry:getProvider(file)
         table.insert(radio_buttons, {
             {
-                text = provider.provider_name,
-                checked = DocumentRegistry:getProvider(file) == provider,
+                text = provider.provider_name .. _(" ~Unsupported"),
```
[the remark this answers is not in the document]
Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: can you explain me how the Template / gettext works for strings (or point me to the docs).
i could not figure it out and keep on getting it wrong.
thanks

File: frontend/apps/filemanager/filemanager.lua (hunk 1 above)
Comment (yparitcher) [reply frontend/apps/filemanager/filemanager.lua]: as of now it only showed  `open with` when there was 2 or more providers, i am showing it also when there are 0 providers, i did not change the behavior for 1 provider (don't show), should i do different?

**Comment by yparitcher** [comment]
adds reset default global provider

FileManager_2019-Oct-17_150651

File: frontend/apps/filemanager/filemanager.lua (hunk 1 above)
Comment (Frenzie) [reply frontend/apps/filemanager/filemanager.lua]: If you allow opening any file as if it were plain text it seems somewhat inconsistent to me to restrict it to unknown filetypes; not quite sure how much sense it makes one way or the other.

Comment (yparitcher) [reply frontend/apps/filemanager/filemanager.lua]: to be more precise i am opening an unknown file with the provider for txt. (crengine). 
if it actually an epub for example crengine will treat it as an epub, if it is unstable crengine will probably crash.

if a file is only supported by one provider there is no reason to try opening it with a different one that probably wont work, and i only am adding this provider if there are no providers already so you will get a `open with` with the same option as opening it directly

Comment (Frenzie) [reply frontend/apps/filemanager/filemanager.lua]: Right, fair enough. Carry on then. ;-)

File: frontend/ui/widget/filechooser.lua [hunk 3]
```diff
@@ -415,6 +426,15 @@ function FileChooser:showSetProviderButtons(file, filemanager_instance, reader_u
                 UIManager:close(self.set_provider_dialog)
             end,
         },
+        {
+            text = _("Reset Default"),
```
Comment (Frenzie) [inline frontend/ui/widget/filechooser.lua]:
```suggestion
            text = _("Reset default"),
```

File: frontend/document/documentregistry.lua [hunk 4]
```diff
@@ -56,9 +56,17 @@ end
...
-    if self.filetype_provider[filename_suffix] then
+    local filetype_provider = G_reader_settings:readSetting("provider") or {}
+    if self.filetype_provider[filename_suffix] or filetype_provider[filename_suffix] then
         return true
     end
+    local DocSettings = require("docsettings")
+    if DocSettings:hasSidecarFile(file) then
+        local doc_settings_provider = DocSettings:open(file):readSetting("provider")
+        if doc_settings_provider then
+            return true
```
Comment (poire-z) [inline frontend/document/documentregistry.lua]: Was going to suggest you should properly do a `docsetting:close()` like it is done in many other places (eg. https://github.com/koreader/koreader/pull/5384).
But no, letting the table disappear when out of scope, as done here, is the right thing to do.
These other places, by calling `:close()`, do unnecessarily flush/save the unmodified data to disk...

File: frontend/ui/widget/filechooser.lua (hunk 3 above)
Comment (poire-z) [reply frontend/ui/widget/filechooser.lua]: Might want to have the buttons on 2 lines, eg.:
<kbd>image</kbd>

File: frontend/ui/widget/filechooser.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (Frenzie) [reply frontend/ui/widget/filechooser.lua]: gettext isn't involved, but the docs I wrote for gettext should include some examples with it.

tl;dr Concatenation like this removes a translator's ability to freely position a word like unsupported.

It's explained [over here](https://github.com/koreader/koreader/blob/7de022a61742f42224a046b604202e3a8e150ed7/doc/Building.md#translations) in a part of the README that probably should be moved elsewhere. (Pinging @pazos.) It's also explained [here](http://koreader.rocks/doc/modules/ffi.util.html#template) but obviously that file is still a bit of a stub documentation-wise.

```
-- @translators %1 is the provider name, in this case Cool Reader Engine
T(_("%1 ~Unsupported"), provider_name)
```

That being said, unless one can also force MuPDF it actually seems unnecessarily complex to use the template function here.

[the remark this answers is not in the document]
Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: i did not want to hardcode a specific provider, to make it future proof

File: frontend/ui/widget/filechooser.lua (hunk 3 above)
Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: i find two lines harder to press, personally

Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: also that would require reworking the whole `open as` widget which i am trying to leave as intact as possible.

File: frontend/ui/widget/filechooser.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: also the word unsupported was a placeholder is there maybe a better word to convey the intent?

**Comment by yparitcher** [comment]
can somone test and make sure i didn't mess up `DocumentRegistry` too much.
thanks

File: frontend/ui/widget/filechooser.lua (hunk 3 above)
Comment (poire-z) [reply frontend/ui/widget/filechooser.lua]: > i find two lines harder to press, personally

Yes, me too. See https://github.com/koreader/koreader/issues/5449#issuecomment-537935889. And we should try to get that better.

But for consistency, and because, if these 3 text fit without truncation in english, they may get truncated in german or in some other language, so it should be on 3 lines.

> also that would require reworking the whole 

No, it should be easy.
Instead of 
```lua
table.insert(buttons, { -- line 1
 {button1}, -- columns
 {button2},
 {button3},
})
```
use
```lua
table.insert(buttons,
 { -- line1
  {button1}, -- columns
  {button2},
 },
 { -- line1
  {button3},
 }
)
```

(Not familiar enough with DocumentRegistry to comment more on this PR.)

Comment (poire-z) [reply frontend/ui/widget/filechooser.lua]: Or better, if you can detect the current value is different than the default, just add that 2nd row with the `Use default` when it is - so in normal conditions, we only see a single row with the 2 buttons.
```lua
table.insert(buttons, { -- line 1
 {button1}, -- columns
 {button2},
})
if there_is_setting_other_than_default then
    table.insert(buttons, { -- line 2
     {button3},
    })
end
```

Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: fixed

**Comment by yparitcher** [comment]
FileManager_2019-Oct-18_140005
FileManager_2019-Oct-18_140020

File: frontend/ui/widget/filechooser.lua [hunk 5]
```diff
@@ -394,18 +394,30 @@ function FileChooser:showSetProviderButtons(file, filemanager_instance, reader_u
...
+                },
+            })
+        end
+    else
+        local provider = DocumentRegistry:getProvider(file)
         table.insert(radio_buttons, {
             {
-                text = provider.provider_name,
-                checked = DocumentRegistry:getProvider(file) == provider,
+                text = provider.provider_name .. _(" ~Unsupported"),
```
Comment (Frenzie) [inline frontend/ui/widget/filechooser.lua]: It should still read something like:
```suggestion
                T(_("%1 ~Unsupported"), provider.provider_name),
```

Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: sorry i cahnged the wrong one last time. Fixed

File: frontend/ui/widget/filechooser.lua [hunk 6]
```diff
@@ -394,18 +394,30 @@ function FileChooser:showSetProviderButtons(file, filemanager_instance, reader_u
...
             {
-                text = provider.provider_name,
-                checked = DocumentRegistry:getProvider(file) == provider,
+                -- @translators %1 is the provider name, such as Cool Reader Engine or MuPDF.
+                text = T(_("%1 ~Unsupported"), provider.provider_name),
+                checked = true,
                 provider = provider,
             },
-        })
+         })
```
Comment (Frenzie) [inline frontend/ui/widget/filechooser.lua]:
```suggestion
        })
```

Comment (Frenzie) [reply frontend/ui/widget/filechooser.lua]: Looks OK to me besides this tiny nitpick, were you done with it?

Comment (yparitcher) [reply frontend/ui/widget/filechooser.lua]: yes

**Comment by Frenzie** [comment]
Alright, thanks!
